core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -74,16 +73,20 @@
         this.unassignedInfo = unassignedInfo;
         this.allocationId = allocationId;
         this.expectedShardSize = expectedShardSize;
+        this.recoverySource = recoverySource;
         assert expectedShardSize == UNAVAILABLE_EXPECTED_SHARD_SIZE || state == ShardRoutingState.INITIALIZING || state == ShardRoutingState.RELOCATING : expectedShardSize + " state: " + state;
         assert expectedShardSize >= 0 || state != ShardRoutingState.INITIALIZING || state != ShardRoutingState.RELOCATING : expectedShardSize + " state: " + state;
         assert !(state == ShardRoutingState.UNASSIGNED && unassignedInfo == null) : "unassigned shard must be created with meta";
+        assert (state == ShardRoutingState.UNASSIGNED || state == ShardRoutingState.INITIALIZING) == (recoverySource != null) : "recovery source only available on unassigned or initializing shard ";
```

Comment:
I don't know what's weird about it. It's actually a good thing that we know already for an unassigned shard where it is to be recovered from (same as we have the restore source on the unassigned shard). This helps for example for the shard explain API to see why a shard cannot be assigned.
We could keep the recovery source when a shard is started but it's pretty much irrelevant at that point for debugging purposes. In the same line of reasoning we could keep the unassigned info around (which we also throw away once shard moves to started).